github_release module get asset_url for latest release or specfic release

srizzling opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the plugin/module/task -->
github_releases

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```

```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

This change adds a feature to the github_releases. I'm still working on this but thought I would open a pull request to ensure that the change is worth being merged upstream. 

Sometimes certain opensource applications that share their code on github do not have there packages available (or at all) in specfic operating system package manager repos (apt, yum, pacman) and sometimes the most reliable way to get the latest provide binaries, tars,  rpms are on github. Sometimes its hard to get updates, when you need them.

This change adds the feature of being to download the latest release without having to script out and expand it into multiple tasks. (get latest release version, wget tar)

It works by calling the releases api, find the latest release, and then finding all the associated assets for that specfic release.

I personally feel like this isn't a a clean way to express it within this task. (should be similar to a package manager action=install, state=latest) but a change like that requires me to edit the "API" which will probs break a few modules. I could also move this to new a module, or not contribute upstream at all :)

I'll cc the main contributor to this script so he can see  if it fits :)
\cc @adrianmoisey

srizzling commented:
And don't worry.. I'll squash all my commits once I'm ready for review.

gregdek commented:
This PR is a work-in-progress. @srizzling, when you're ready to have this PR reviewed, please remove WIP from the title of the PR.

sivel commented:
My opinion here is that I would prefer this module or a new module allow the URL(s) of the latest release to be returned.  Then you could utilize something like `get_url` to perform the fetching.

srizzling commented:
Thats fair! Then we can also use the unarchive module (or even the package manager module). Will see I have time to tackle this today.

srizzling commented:
ready_for_review

srizzling commented:
Hey,

So as @sivel suggested I have changed the functionality a bit and squashed into one commit to make it easier to review.

The action get_asset_url accepts an asset_regex and a release_version. If the release_version has not been provided it default to latest. This allows you to lock the version and maintain Idempotence.

I'm pretty fresh to python, so please excuse the uglyness :)

Thanks,
Sriram

gregdek commented:
Thanks @srizzling for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

srizzling commented:
ready_for_review

adrianmoisey commented on source_control/github_release.py:
```diff
@@ -16,6 +16,8 @@
 # You should have received a copy of the GNU General Public License
 # along with Ansible. If not, see <http://www.gnu.org/licenses/>.
 
+import tempfile
```
Since the refactor, is this still needed?

gregdek commented:
Thanks @srizzling. To the current maintainers, @adrianmoisey please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

srizzling commented:
ready_for_review
